contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+      for (uint256 i = 0; i < ids.length; i++) {
+        uint256 id = ids[i];
+        if (amounts[i] > claimableBalanceForScheduleForAccount(id, from)) {
+          revert InsufficientClaimableBalance({account: from, scheduleId: id});
+        }
+      }
+    }
+    return super._beforeTokenTransfer(operator, from, to, ids, amounts, data);
+  }
+
+  /**
+   * Linearly released balance for a single schedule at the current block timestamp, ignoring any
+   * released amount floor that has been set for the schedule.
+   */
+  /* solhint-disable not-rely-on-time, this is time-dependent */
+  function _linearReleaseAmountAvailable(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 linearAmountAvailable;
+    // slither-disable-next-line timestamp this contract is block time dependent
+
+    if (block.timestamp >= schedule.endTime) {
+      linearAmountAvailable = totalSupply(scheduleId);
+    } else {
+      uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+      // slither-disable-next-line timestamp this contract is block time dependent
+      linearAmountAvailable = block.timestamp < schedule.startTime
+        ? 0
+        : (_scheduleTrueTotal(schedule, scheduleId) *
+          (block.timestamp - schedule.startTime)) / rampTotalTime;
+    }
+    return linearAmountAvailable;
+  }
+
+  /* solhint-enable not-rely-on-time */
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
+    uint256 scheduleId,
+    address account
+  ) private view returns (uint256) {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 scheduleTrueTotal = _scheduleTrueTotal(schedule, scheduleId);
+    uint256 revocableForAccount;
+    // avoid division by 0
+    if (scheduleTrueTotal == 0) {
+      revocableForAccount = 0;
+    } else {
+      uint256 balanceOfAccount = balanceOf(account, scheduleId);
+      uint256 claimedAmountForAccount = schedule.claimedAmountsByAddress[
+        account
+      ];
+      revocableForAccount =
+        ((claimedAmountForAccount + balanceOfAccount) *
+          (totalQuantityToRevoke)) /
+        scheduleTrueTotal;
+    }
+
+    return revocableForAccount;
+  }
+
+  /**
+   * @dev The total amount of released tokens available at the current block timestamp for the schedule.
+   * Takes the maximum of either the calculated linearly released amount based on the schedule parameters,
+   * or the released amount floor, which is set at the current released amount whenever the balance of a
+   * schedule is decreased through revocation or withdrawal.
+   */
+  function _releasedBalanceOfSingleSchedule(uint256 scheduleId)
```

Comment:
it looks like in every instance you call this function you load are calling it like this

```sol
Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
_releasedBalanceOfSingleSchedule(schedule.id);
```
but then this function loads schedule again on line 1015 resulting in 2 loads of the schedule struct to storage

why not just pass what you need from the schedule when you load it the first time?

Reply:
there's also a place where this is called in a context that doesn't already have a schedule loaded from memory, but there are two places where it's called and it does, so I'll refactor to remove the two loads and add the one.